### PR TITLE
Fix commutativity of `Hamiltonian` and `Observable` in tensor product

### PR DESCRIPTION
**Pull request by Jaybsoni**
**Context:**
Given a Hamiltonian object `H` and an observable object (eg. `qml.PauliX(0)`), the tensor product: 

```python 
dev = qml.device("default.qubit", wires=3)

@qml.qnode(dev)
def circ(op):
    qml.Hadamard(0)
    qml.CNOT(wires=[0,1]) 
    return qml.expval(op)
```

```
>>> op1 = H @ qml.PauliX(0)
>>> op2 = qml.PauliX(0) @ H   
>>>
>>> circ(op1)  # works as expected 
>>> circ(op2) # fails due to an error accessing the EigenValues of a tensor observable
```

**Description of the Change:**
The `__matmul__` method in the `Observable` class checks if the other object is an instance of `Hamiltonian`. In this case it will differ the matrix multiplication to the `Hamiltonian` classes implementation. Also a `__rmatmul__` method is implemented for the `Hamiltonian` class to allow for the above. 

**Related GitHub Issues:**
User reported [bug](https://github.com/PennyLaneAI/pennylane/issues/2557)

### CODE REVIEW

**Comment by Jaybsoni** [comment]
[sc-19496]

**Review by dime10 — APPROVED** [review]
Thanks @Jaybsoni, this looks great! 💯 

Just need to add a changelog entry.

**Comment by Jaybsoni** [comment]
codecov :/

File: pennylane/ops/qubit/hamiltonian.py [hunk 1]
```diff
@@ -544,6 +544,23 @@ def __matmul__(self, H):
 
         raise ValueError(f"Cannot tensor product Hamiltonian and {type(H)}")
 
+    def __rmatmul__(self, H):
+        r"""The tensor product operation from the right) between a Hamiltonian and
```
Comment (Jaybsoni) [inline pennylane/ops/qubit/hamiltonian.py:548]:
```suggestion
        r"""The tensor product operation (from the right) between a Hamiltonian and
```

**Comment by antalszava** [comment]
Just a fly-by name change of the PR to showcase that we'd like to have the commutative property and it was fixed (any comparable other PR name works great too).

**Comment by Jaybsoni** [comment]
@antalszava, the codecov issue seems to be for another file, can we override this and merge?

**Comment by Jaybsoni** [comment]
Thanks !

**Pull request merged**